feat(release post): add v1.27 release blog post. (#355)
kubewarden/kubewarden.io: content/blog/2025/07/kubewarden-1.27-release.md

  apiGroup: rbac.authorization.k8s.io
```

Let's try to deploy something:

```console
kubectl create -f - << EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: super-admin-deployment
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: super-admin-app
  template:
    metadata:
      labels:
        app: super-admin-app
    spec:
      serviceAccountName: super-admin-sa
      containers:
        - name: main
          image: nginx:alpine
          ports:
            - containerPort: 80
        - name: kubectl
          image: bitnami/kubectl:latest
          command: ["sleep", "infinity"]
EOF
Error from server: error when creating "STDIN": admission webhook "clusterwide-high-risk-service-account.kubewarden.admission" denied the request: Cannot use service account 'system:serviceaccount:default:super-admin-sa' with permissions to perform list /secrets in the cluster
```

## New _kubernetes/can-i_ Host Capability

To make the High-Risk Service Account Blocker policy more efficient, this
release introduces a new host capability:
[`kubernetes/can_i`](https://docs.kubewarden.io/reference/spec/host-capabilities/kubernetes#operation---can_i).

> This capability leverages the Kubernetes `SubjectAccessReview` API to check if
> a service account has permission to perform certain operations. This approach
> avoids the need to fetch numerous RBAC resources to determine if the service
> account can perform the blocked operations, making the check much faster and
> more reliable.

Reviewer: ```suggestion
This capability uses the Kubernetes `SubjectAccessReview` API to check if
a service account has permission to perform certain operations. This approach
avoids the need to fetch numerous RBAC resources to determine if the service
account can perform the blocked operations, making the check faster and
more reliable.
```

> To support this new capability, the default `ServiceAccount` used in the Helm
> chart to run the Policy Server has also been updated to add the permission to
> create `SubjectAccessReview` resources. If you want to use this new capability
> with a Policy Server that uses a different `ServiceAccount`, you must update
> its permissions accordingly.

Reviewer: ```suggestion
To support this new capability, the default `ServiceAccount` used in the Helm
chart to run the Policy Server has been updated to add permission to
create `SubjectAccessReview` resources. If you want to use this new capability
with a Policy Server that uses a different `ServiceAccount`, you must update
its permissions accordingly.
```

## _kwctl_ Now Mirrors _policy-server_ Behavior

The `kwctl` CLI now performs post-policy processing validations and changes

> that were previously executed only by the `policy-server`. We've moved logic
> for features like custom error messages and mutation permissions from the
> `policy-server` codebase to their common [evaluation
> library](https://github.com/kubewarden/policy-evaluator).

Reviewer 2: ```suggestion
that were previously executed only by the `policy-server`. We've moved the logic
for features like custom error messages and mutation permissions from the
`policy-server` codebase to a shared [evaluation
library](https://github.com/kubewarden/policy-evaluator).
```

> As a result, users running policies with `kwctl` will now see the exact same
> behavior that the `policy-server` would produce. This is great for testing
> policy configurations before deploying them to a cluster.

Reviewer: ```suggestion
As a result, users running policies with `kwctl` now see the same
behavior that the `policy-server` produces. This is great for testing
policy configurations before deploying them to a cluster.
```

> The post-policy steps performed also by `kwctl` are:

Reviewer: ```suggestion
The post-policy steps also performed by `kwctl` are:
```

- policy mode enforcement: when the policy is in monitor mode, `kwctl` returns the request as allowed, even if the request was

>   rejected by the policy. Printing a warning message with the original result

Reviewer: ```suggestion
  rejected by the policy. It prints a warning message with the original result.
```

- custom error message: `kwctl` will return the error message defined in the
  `message` field when the policy is rejected

> - reject mutated requests: `kwctl` will reject a request if the policy mutate

Reviewer 2: ```suggestion
- reject mutated requests: `kwctl` will reject a request if the policy mutates
```

introduced a significant number of new host capabilities and improved the
codebase, paving the way for a public release. With that in mind, he is now
working on an alpha release of the SDK to validate the full release process and
test it with a template policy.

## Policy Fixes

We recently received a
[request](https://github.com/kubewarden/rego-policies-library/issues/50) from
the community to implement an additional validation in one of our Rego
policies. The **Kustomization Target Namespace** policy, which validates FluxCD
Kustomizations, now ensures that the optional `spec.targetNamespace` field is
set to one of the allowed namespaces. Prior to this change, requests were
accepted even if this field was not set.

## SBOM Generation Update and maintenance tasks

In this release, beyond the usual dependency updates and maintenance work,
we've removed all references to the deprecated
[spdx-sbom-generator](https://github.com/opensbom-generator/spdx-sbom-generator)
tool. All our GitHub Actions now use [Syft](https://github.com/anchore/syft) to
generate SBOM files. The generated files are slightly different due to the

> change in tooling, but users should not be concerned, as the files are

Reviewer: ```suggestion
change in tooling, but users should not be concerned, the files are
```

---
title: "Kubewarden 1.27 Release"
authors:
  - José Guilherme Vanz
date: 2025-07-23
---

Here's a look at the key updates and improvements in the latest release.

## New High-Risk Service Account Policy

In this release, we've introduced a new policy to improve cluster security. The
**High-Risk Service Account Blocker** policy, as its name suggests, blocks
workloads that attempt to run with a service account that has excessive
permissions.

> This policy allows cluster operators to define a list of forbidden permissions.

Reviewer 3: I would like to have a small ClusterAdmissionPolicy example of the new policy to showcase it better.

Workloads using a service account with any of these permissions will be
blocked. Check out the policy's
[repository](https://github.com/kubewarden/high-risk-service-account-policy) to
learn more about its configuration.

## New _kubernetes/can-i_ Host Capability

To make the High-Risk Service Account Blocker policy more efficient, this
release introduces a new host capability:
[`kubernetes/can_i`](https://docs.kubewarden.io/reference/spec/host-capabilities/kubernetes#operation---can_i).

This capability leverages the Kubernetes SubjectAccessReview API to check if a
service account has permission to perform certain operations. This approach
avoids the need to fetch numerous RBAC resources to determine if the service
account can perform the blocked operations, making the check much faster and
more reliable.

> ## _kwctl_ Now Mirrors _policy-server_ Behavior

Reviewer 3: We should mention which behaviours are those: policy mode (accepting and printing a warning when on `monitor` mode), honouring the custom rejection message, rejecting and printing an error if a policy mutates a request but doesn't declare it is a mutating policy. It would be nice to have small 2 liner examples showcasing the output.

The `kwctl` CLI now performs post-policy processing validations and changes
that were previously executed only by the `policy-server`. We've moved logic
for features like custom error messages and mutation permissions from the
`policy-server` codebase their common [evaluation
library](https://github.com/kubewarden/policy-evaluator).

As a result, users running policies with `kwctl` will now see the exact same
behavior that the `policy-server` would produce. This is great for testing
policy configurations before deploying them to a cluster.

## GSoC Update: JavaScript SDK Enhancements

This month also saw great enhancements to our JavaScript SDK. A big thank you

> to **Esosa**, our GSoC mentee!

Reviewer 4: Maybe we can add a link to his GH profile?

Esosa is focused on making the SDK stable and production-ready. He has
introduced a significant number of new host capabilities and improved the
codebase, paving the way for a public release. With that in mind, he is now
working on an alpha release of the SDK to validate the full release process and
test it with a template policy.

## Policy Fixes

We recently received a
[request](https://github.com/kubewarden/rego-policies-library/issues/50) from
the community to implement an additional validation in one of our Rego
policies. The **Kustomization Target Namespace** policy, which validates FluxCD
Kustomizations, now ensures that the optional `spec.targetNamespace` field is
set to one of the allowed namespaces. Prior to this change, requests were
accepted even if this field was not set.

## SBOM Generation Update and maintenance tasks

In this release, beyond the usual dependency updates and maintenance work,
we've removed all references to the deprecated
[spdx-sbom-generator](https://github.com/opensbom-generator/spdx-sbom-generator)

> tool. All our GitHub Actions now use syft to generate SBOM files.

Reviewer 3: I think we should mention there's a change in the resulting SPDX artifact, albeit they are equivalent.
